### DISCUSSION TITLE
Beginner's question

### DISCUSSION THREAD

**Discussion by JanChrisKlein** [peterw8102/Z80-Retro#11] [General]
Hello,

totally new to CP/M and retro computing ... I could use some starter advice...
I've got a PCB V 2.2 fully populated. So far all I have managed to do was flash 'zloader.hex' (Z80-Retro/images) into the ROM.
Nothing on the SDcard. No Raspi connected. Jumpers are set as per default. And all I get is:

Z80 ZIOS 2.1.0
2000-01-01 00:00.00

and a blinking square. Time and date never change. 
The prompt in itself is kind of a success but won't keep me entertained forever.
I can't seem to enter anything here, so I don't see a chance to prepare the SDcard. 
The Raspi, as I understood it, is an alternative way to boot from but I wouldn't know how to do that and I also wouldn't know how to prepare the Raspi for that role.

So, my problems couldn't be more basic, I guess. Any advice?

Christoph

**Comment by peterw8102** [comment]
Hi Christoph, you're definitely getting close if you got the output above.

I've made a few bug fixes since I uploaded the hex file you're using. Could you try the version I've just added to the repo? images/zloader.hex.

When you run that it should report version 2.1.3.

Pete

**Reply by JanChrisKlein** [reply]
Hi Pete,
thanks a lot for Your rapid response. I now have ...

```
Z80 ZIOS 2.1.2
2024-01-03 22:08.05

Cfg Sw                   07 - Console: VDU. Boot ZLoader
SDCard1                  YES
SDCard2                  YES
Video                    NO
PIO                      NO

>
```
... and I can type in commands and set my clock :)

I have also been able to send hex files (like 'cpm22.hex') to the Z80. 
What doesn't work yet is actually starting CP/M:

```
> bos?
ID NAME     DV SDADDR   LOAD  LEN EXEC FL
03 CP/M     00 00002000 DB00 2300 F10C 81
02 CP/M     00 00002000 DB00 2300 F10C 81
> bos 02
Image: CP/M
```
after which nothing happens.
Are the values given in the 'WB' line still valid?

`WB 3 0:A DB00 2300 F10C 01 CP/M
`

Am I missing something else? Are we there yet? ;)

Christoph

**Reply by peterw8102** [reply]
Use:

```
WI 0:A DD00 FE00
```
and

```
WB 3 0:A DD00 2100 F30C 01 CP/M
```
I shaved 512 bytes of the size of the BIOS. Let me know if that works and I'll fix the documentation. I packed away my hardware to make room for Christmas and I haven't set it up again yet so can't test this myself.

**Comment by JanChrisKlein** [comment]
Ooookay ... getting there, thanks again! I now got CP/M going :)
What I am now struggling with is filling my disk:
Under "installing CP/M" under the topic "Loading applications" You are mentioning a file "cpm22apps.pkg". It's supposed to be in the "images" folder, but I can't find it there. Did it move?

And another question: how many disks should there be? When I start CP/M I am on drive "A". I can switch to "B", "C", "D" but when I reach "E" I get "Bdos Err On E: Select". Shouldn't there be 16 drives?

Christoph

**Reply by peterw8102** [reply]
Definitely getting closer :-)

You should find the [cpm22apps.pkg](https://github.com/peterw8102/Z80-Retro/blob/master/images/cpm22apps.pkg) in the images directory now.

You'll also find `download.hex` in the cpmtools directory. That needs to be manually installed before you can use the `.pkg` file. I've [documented how to do that here](https://github.com/peterw8102/Z80-Retro/wiki/Installing-CPM#install-downloadcom).

**Reply by peterw8102** [reply]
Second part of your question. CP/M supports up to 16 drives. Each drive however increases the size of the BIOS. Given I'm using 4MB virtual disks on the SDCard I decided for now to limit the drives from A-D. You can change which of the virtual disks is mounted on which A-D drive letter using the 'mount.com' programme which is in the package file.

If you mount a disk that's not been used before it will need formatting - so see the 'format.com' programme.

By default CP/M boots with drive A mounted with the drive on which you installed CP/M - the one you formatted and used in my "installing CP/M' instructions.

**Reply by JanChrisKlein** [reply]
I've downloaded the .pkg-file and installed it as per instructions. Worked well, lots of CP/M tools now on "A:".
One small thing: There is one file without name and extension which doesn't seem right ... ?
Don't know how it got there and I don't know how to remove it.

Christoph

**Reply by peterw8102** [reply]
Glad you got there, I've been updating the documentation to make it easier for others in future.

I don't know where that 'no name' file has come from. What do you get when you run `dir`, `dirx` and `stat *.*`?

**Reply by JanChrisKlein** [reply]
Well, I do occasionally screw up ;)
I've done the whole procedure more than one time and I was not always successful...
(by the way: is there a _definitive_ way to bring an SDcard back to its factory state?)

I've noticed an error while downloading 'cpm22apps.pkg':

```
a>download
................................
................................
|
======File Length Error======

a>59
59?

a>A:DOWNLOAD CR.COM
................................
................|
OK
```
(always at the very start).
Right now the download doesn't seem to work at all
(LED1 and the LED on the SDCard module are lighting up permanently).

Christoph

**Reply by JanChrisKlein** [reply]
I found out how to get rid of my "noname" file (easy, really:  'ERA *. ', with a space as the 
file extension).

Chris

**Reply by peterw8102** [reply]
It looks like you're typing the command `download` on the command line then pasting the contents of the pkg file. Don't enter the command! Just paste the contents to the file at the `a>` prompt.

This is what I see:
```
a>save 4 download.com
a>dir
A: DOWNLOAD COM
a>A:DOWNLOAD ASM.COM
................................
................................
|
OK

a>
a>A:DOWNLOAD CR.COM
................................
................|
OK

a>
```

**Comment by JanChrisKlein** [comment]
Oh, and there is still a little quirk in Your instructions, I think:

> Once you have all this information you can add an entry to the boot menu with:

`WB 3 0:A DD00 2300 F30C 01 CP/M`

should be

`WB 3 0:A DD00 2100 F30C 01 CP/M`

shouldn't it?

Also, setting the image to '3' means that booting into CP/M directly won't work,
because ZLOADER seems to be looking for the CP/M installation at image '1' (?).
(easily remedied by writing 'WB 1 0:A ...', of course.

Christoph

**Reply by peterw8102** [reply]
If you use a number other than 1 then you can 'boot' that system using `BOS n`, in the example above `BOS 3`.

An ID of 1 is useful if you want the system to automagically boot into, say, CP/M. In this case [setting the DIP switches](ZLoader-Configuration-Switches) to 'Boot OS image 1 from the SDCard'. I tend to have a set of different boot images for different CP/M configurations. I've played with using alternative CCPs for example.

**Comment by JanChrisKlein** [comment]
```
dirx
Name    Ext Bytes   Name    Ext Bytes   Name    Ext Bytes   Name    Ext Bytes
              10K ! ED      COM    8K ! MBASIC  COM   24K ! SUBMIT  COM    2K
CR      COM    6K ! FORMAT  COM    2K ! MOUNT   COM    2K ! UNCR    COM   10K
DDT     COM    6K ! LBREXT  COM   10K ! PIP     COM    8K ! UNZIP   COM    4K
DDTZ    DOC   30K ! LIB     COM    8K ! PMARC   COM   12K ! XSUB    COM    2K
DIRX    COM    4K ! LINK    COM   16K ! PMEXT   COM   14K ! ZSID    COM   10K
DOWNLOADCOM    2K ! LOAD    COM    2K ! RMAC    COM   14K
DUMP    COM    2K ! MAC     COM   12K ! STAT    COM    6K
26 File(s), occupying 226K of 4016K total capacity
484 directory entries and 3790K bytes remain on A:
a>
```
This happened with both of my boards, b.t.w. ...
Chris

**Comment by JanChrisKlein** [comment]
> I've downloaded the .pkg-file and installed it as per instructions. Worked well, lots of CP/M tools now on "A:". One small thing: There is one file without name and extension which doesn't seem right ... ? Don't know how it got there and I don't know how to remove it.
> 
> Christoph
